crates/flowctl/src/controlplane.rs

```diff
@@ -0,0 +1,95 @@
+use std::ops::Deref;
+use std::sync::Arc;
+
+/// A wafer-thin wrapper around a `Postgrest` client that makes it
+/// cheaply cloneable and implements `Debug`. This allows us to create
+/// a client once and then store it in the `CliContext` for future re-use.
+/// This client implements `Deref<Target=Postgrest>`, so you can use it
+/// just like you would the normal `Postgrest` client.
+#[derive(Clone)]
+pub struct Client(Arc<postgrest::Postgrest>);
+
+impl Debug for Client {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // We can't really give a better debug impl since Postgrest
+        // keeps all of its members private.
+        f.write_str("controlplane::Client")
+    }
+}
+
+impl Deref for Client {
+    type Target = postgrest::Postgrest;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+/// Creates a new client. **you should instead call `CliContext::controlplane_client(&mut Self)`**, which
+/// will re-use the existing client if possible.
+pub(crate) fn new_client(config: &Config) -> anyhow::Result<Client> {
+    match &config.api {
+        Some(api) => {
+            // Try to give users a more friendly error message if we know their credentials are expired.
+            check_access_token(&api.access_token)?;
+            let client = postgrest::Postgrest::new(api.endpoint.as_str());
+            let client = client.insert_header("apikey", &api.public_token);
+            let client =
+                client.insert_header("Authorization", format!("Bearer {}", &api.access_token));
+            Ok(Client(Arc::new(client)))
+        }
+        None => {
+            anyhow::bail!("You must run `auth login` first")
+        }
+    }
+}
+
+pub fn configure_new_access_token(ctx: &mut CliContext, token: String) -> anyhow::Result<()> {
+    // try to catch issues caused by missing or extra data that may have been accidentally copied
+    let email = check_access_token(&token)?;
+    ctx.config_mut().set_access_token(token);
+    println!("Configured access token for user: '{email}'");
+    Ok(())
+}
+
+fn check_access_token(token: &str) -> anyhow::Result<String> {
+    let jwt = parse_jwt(token).context("invalid access_token")?;
+    // Try to give users a more friendly error message if we know their credentials are expired.
+    if jwt.is_expired() {
+        anyhow::bail!("access token is expired, please re-authenticate and then try again");
+    }
+    Ok(jwt.email)
+}
+
+#[derive(Deserialize)]
+struct JWT {
+    email: String,
+    exp: i64,
+}
+
+impl JWT {
+    fn is_expired(&self) -> bool {
+        let exp = time::OffsetDateTime::from_unix_timestamp(self.exp).unwrap_or_else(|err| {
+            tracing::error!(exp = self.exp, error = %err, "invalid exp in JWT");
+            time::OffsetDateTime::UNIX_EPOCH
+        });
+        time::OffsetDateTime::now_utc() >= exp
+    }
+}
+
+fn parse_jwt(jwt: &str) -> anyhow::Result<JWT> {
+    let payload = jwt
```

Comment:
I was going to suggest using a library for decoding JWTs and also verifying the signature at the same time, but it looks like Supabase actually doesn't support public/private keys for their JWTs, so it doesn't seem worth it. https://github.com/supabase/supabase/discussions/5911#discussioncomment-2371302

Reply:
yea, this is a little bit unfortunate, but seemed like the best thing in the short term, especially given our future plans of having more of a legit API key.